BeeSwift/SignInView.swift

```diff
@@ -0,0 +1,130 @@
+                            .textFieldStyle(RoundedBorderTextFieldStyle())
+                            .textInputAutocapitalization(.never)
+                            .submitLabel(.done)
+                        
+                        Button(action: signIn) {
+                            Text("Sign In")
+                                .font(.system(size: 20))
+                                .foregroundColor(.white)
+                                .frame(maxWidth: .infinity)
+                                .frame(height: 44)
+                        }
+                        .background(Color.gray)
+                        .cornerRadius(8)
+                        .disabled(viewModel.isLoading)
+                    }
+                    .frame(maxWidth: .infinity)
+                    .padding(.horizontal, 40)
+                    
+                    Spacer()
+                    Spacer()
+                    Spacer()
+                }
+                .padding()
+                
+            }
+            .alert("Could not sign in", isPresented: $showingFailedSignInAlert) {
+                Button("OK", role: .cancel) { }
+            } message: {
+                Text("Invalid credentials")
+            }
+            .alert("Incomplete Account Details", isPresented: $showingMissingDataAlert) {
+                Button("OK", role: .cancel) { }
+            } message: {
+                Text("Username and Password are required")
+            }
+            .onReceive(NotificationCenter.default.publisher(for: Notification.Name(CurrentUserManager.failedSignInNotificationName))) { _ in
+                viewModel.isLoading = false
+                showingFailedSignInAlert = true
+            }
+            .onReceive(NotificationCenter.default.publisher(for: Notification.Name(CurrentUserManager.signedInNotificationName))) { _ in
+                viewModel.isLoading = false
+                dismiss()
+            }
+        }
+        .overlay {
+            if viewModel.isLoading {
+                ProgressView()
+                    .scaleEffect(1.5)
+                    .frame(maxWidth: .infinity, maxHeight: .infinity)
+                    .background(Color.black.opacity(0.2))
+            }
+        }
+    }
+    
+    private func signIn() {
+        guard !viewModel.email.trimmingCharacters(in: .whitespaces).isEmpty,
+              !viewModel.password.isEmpty else {
+            showingMissingDataAlert = true
+            return
+        }
+        
+        Task {
+            await viewModel.signIn()
+        }
+    }
+}
+
+@MainActor
+class SignInViewModel: ObservableObject {
+    @Published var email = ""
+    @Published var password = ""
+    @Published var isLoading = false
+    
+    func signIn() async {
+        isLoading = true
+        
+        do {
+            try await Task.sleep(nanoseconds: 3_000_000_000)
+        } catch {
+            print("canceled early")
+        }
```

Comment:
The min process time was intentional so as to provide for a better user experience, actually. When tapping the button to trigger a sign in, it is usually a very quick process and one hardly has a chance to notice the progress indicator or the network activity indicator. If we are to show these, then they should be shown long enough for the user to have had a chance to recognize them.

Reply:
Plus when the user logs in, the endpoint returns a user object completed with a list of goal names. It could be that the app then presents a loading state of the gallery with the data it has while it continues fetching each goal and filling in the cells.

Reply:
I could _maybe_ get behind this argument. However, I also log in/out a lot as part of testing the app, and waiting 3 seconds every time would be a significant frustration.

I'd be open to having the login screen continue to show in the loading state until all goals are loaded into the gallery controller. But no explicit additional delays.